[FLINK-15487][table] Allow registering FLIP-65 functions in TableEnvironment

twalthr opened the pull request:
## What is the purpose of the change

This updates all catalog related interfaces to FLIP-65. It also continues FLIP-64 by exposing new interfaces in table environment for registering functions of temporary/system/catalog kind. Furthermore, it adds early class validation to the function catalog.

The first functions will be fully functional once the code generator has been updated.

## Brief change log

- Update of `TableEnvironment`
- Update of `FunctionCatalog`
- Update of `FunctionCatalogOperatorTable`

## Verifying this change

- New `FunctionCatalogTest` cases
- New `UserDefinedFunctionHelperTest` case

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

dawidwys commented on flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/functions/UserDefinedFunctionHelperTest.java:
```diff
@@ -0,0 +1,293 @@
+	// --------------------------------------------------------------------------------------------
+
+	/**
+	 * Test specification shared with the Scala tests.
```
nit: Is this comment valid?

twalthr replied:
nope copy paste error

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java:
```diff
@@ -81,8 +85,283 @@ public void setPlannerTypeInferenceUtil(PlannerTypeInferenceUtil plannerTypeInfe
+	 * Registers a temporary catalog function.
+	 */
+	public void registerTemporaryCatalogFunction(
+			UnresolvedIdentifier unresolvedIdentifier,
```
We're not being too consistent with the type of identifier that the `FunctionCatalog/CatalogManager` accepts (I'm guilty of that myself).

Some methods accept `UnresolvedIdentifier` e.g. `FunctionCatalog#registerTemporaryCatalogFunction`, `CatalogManager#dropTemporaryView`. Some resolved `ObjectIdentifier` e.g. `CatalogManager#createTemporaryTable`, `CatalogManager#createTable`.

I am not sure which one should we prefer. If we go with the `UnresolvedIdentifier` the benefit is that we always qualify in the `Catalog*`. The downside is that we would no longer qualify in the `*Operations`, (e.g. CreateTableOperation etc.), whereas we said that all Operations should be fully resolved...

We can create a follow up for that.

twalthr replied:
I also observed these inconsistencies. I think the catalog manager can resolve identifiers internally and thus can just expose `UnresolvedIdentifier`. That's why all my newly added methods take the unresolved one.

We said that operation should be fully resolved but that was before the DDL. So things might have changed a bit. Maybe we should distinguish between resolved and unresolved operations. Let's discuss this in a follow up issue.

dawidwys replied:
Created https://issues.apache.org/jira/browse/FLINK-15859

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:
```diff
@@ -273,6 +274,72 @@ public void registerFunction(String name, ScalarFunction function) {
 
+	@Override
+	public void createTemporarySystemFunction(String name, Class<? extends UserDefinedFunction> functionClass) {
+		final UserDefinedFunction functionInstance = UserDefinedFunctionHelper.instantiateFunction(functionClass);
```
Yes, that makes sense. We can do this while integrating the `FunctionCatalog` into `CatalogManager`.

dawidwys replied:
Created https://issues.apache.org/jira/browse/FLINK-15860

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java:
```diff
@@ -97,7 +97,13 @@
-			final FunctionDefinition definition = prepareUserDefinedFunction(unresolvedCall.getFunctionDefinition());
+			final FunctionDefinition definition;
+			// clean functions that were not registered in a catalog
+			if (!unresolvedCall.getFunctionIdentifier().isPresent()) {
```
Yes, this is a shortcoming. I will create a follow up issue for this.

twalthr replied:
https://issues.apache.org/jira/browse/FLINK-15804

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java:
```diff
@@ -238,36 +244,40 @@ private ResolvedExpression runLegacyTypeInference(
 					taf.getResultTypeInfo(),
 					taf.getAccumulatorTypeInfo());
+			} else if (definition instanceof UserDefinedFunction) {
+				UserDefinedFunctionHelper.prepareInstance(
```
Not yet :(

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java:
```diff
@@ -147,39 +157,156 @@
+	 */
+	private static void validateImplementationMethods(Class<? extends UserDefinedFunction> functionClass) {
+		if (ScalarFunction.class.isAssignableFrom(functionClass)) {
+			validateImplementationMethod(functionClass, false, false, "eval");
```
I think this helper class is a good location. Will update the PR.

twalthr commented:
@flinkbot run travis

twalthr commented:
@dawidwys I addressed your feedback and Travis gives green light. If you have no objections, I will merge this. Can you take care of opening follow up issues? Thanks.
